src/librustdoc/html/static/js/search.js

```diff
@@ -127,13 +118,196 @@ window.initSearch = function(rawSearchIndex) {
 
     /**
      * Executes the query and builds an index of results
-     * @param  {[Object]} query      [The user query]
-     * @param  {[type]} searchWords  [The list of search words to query
-     *                                against]
-     * @param  {[type]} filterCrates [Crate to search in if defined]
-     * @return {[type]}              [A search index of results]
+     * @param  {[string]} val      [The user query]
+     * @return {[Object]}          [The parsed query]
      */
-    function execQuery(query, searchWords, filterCrates) {
+    function parseQuery(val) {
+        function isWhitespace(c) {
+            return " \t\n\r".indexOf(c) !== -1;
+        }
+        function isSpecialStartCharacter(c) {
+            return "(<\"".indexOf(c) !== -1;
+        }
+        function isStopCharacter(c) {
+            return isWhitespace(c) || "),>-=".indexOf(c) !== -1;
+        }
+        function getStringElem(query) {
+            query.pos += 1;
+            while (query.pos < query.length && query.val[query.pos] !== "\"") {
+                if (query.val[query.pos] === "\\") {
+                    // We ignore the next coming character.
+                    query.pos += 1;
+                }
+                query.pos += 1;
+            }
+            // To skip the quote at the end.
+            query.pos += 1;
+        }
+        function skipWhitespaces(query) {
```

Comment:
They're specific to the parser and shouldn't be used in another context, which is why I moved them all in the `parseQuery` function. I don't want them to be accessible.

Reply:
We talked about this on Discord recently: https://discord.com/channels/442252698964721669/459149231702278154/920563608285487134, and I thought you were on board with the idea that these kinds of functions should be moved to a higher level.

I even posted a PR for some refactoring along these lines (https://github.com/rust-lang/rust/pull/91992).

In my opinion, nesting lots of functions like this is a very bad readability problem for this code. It means that when you're looking at the definition of a function (like `parseQuery`), you can't see where its body is implemented. And even scrolling down to find the body, it's very hard to keep track of where the inner functions end.

The right approach for this type of scope isolation in Rust is a `mod`. In JS it's a module, or at a minimum a separate file with an [Immediately-Invoked Function Expression (IIFE)](https://developer.mozilla.org/en-US/docs/Glossary/IIFE): `(function() {})()`.

So if it's important that these helper functions not be visible in the scope of the broader search.js, what I recommend is splitting out a `parse-query.js`, wrap it in an IIFE, and at doc build time we can concatenate `search.js` and `parse-query.js`

Alternately we could define a `class QueryParser` and provide private methods on it, with a single public method `parseQuery`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields.

Reply:
That sounds like a good idea to a `QueryParser` class. Considering this PR is already huge, do you mind if I do it in a follow-up PR?

Reply:
The goal of the refactoring to have QueryParser is to make the code easier to read. I'm still struggling a bit to fully digest this PR, and one of the obstacles is the nested functions I mention here. I think if you can do this refactoring as part of this PR, it would make it much easier for me to do a good job finishing the review.

Reply:
Ok, I'll make it into a separate commit.